src/number_affixes.zig

```diff
@@ -0,0 +1,150 @@
+const std = @import("std");
+const mem = std.mem;
+
+pub const Prefix = enum {
+    binary,
+    octal,
+    decimal,
+    hex,
+
+    pub fn digitAllowed(prefix: Prefix, c: u8) bool {
+        return switch (c) {
+            '0', '1' => true,
+            '2'...'7' => prefix != .binary,
+            '8'...'9' => prefix == .decimal or prefix == .hex,
+            'a'...'f', 'A'...'F' => prefix == .hex,
+            else => false,
+        };
+    }
+
+    pub fn fromString(buf: []const u8) Prefix {
+        if (buf.len == 1) return .decimal;
+        // tokenizer enforces that first byte is a decimal digit or period
+        switch (buf[0]) {
+            '.', '1'...'9' => return .decimal,
+            '0' => {},
+            else => unreachable,
+        }
+        switch (buf[1]) {
+            'x', 'X' => return if (buf.len == 2) .decimal else .hex,
+            'b', 'B' => return if (buf.len == 2) .decimal else .binary,
+            else => {
+                if (mem.indexOfAny(u8, buf, "eE.")) |_| {
+                    // This is a decimal floating point number that happens to start with zero
+                    return .decimal;
+                } else if (Suffix.fromString(buf[1..], .int)) |_| {
+                    // This is `0` with a valid suffix
+                    return .decimal;
+                } else {
+                    return .octal;
+                }
+            },
+        }
+    }
+
+    pub fn radix(prefix: Prefix) u8 {
+        return switch (prefix) {
+            .binary => 2,
+            .octal => 8,
+            .decimal => 10,
+            .hex => 16,
+        };
+    }
```

Comment:
`Prefix` enum tags should have these values and calls to this function should be replaced with `@enumToInt`.